primo2/inference/dynamic.py

```diff
@@ -0,0 +1,305 @@
+
+    Inference is exact and based on factor trees. The marginal posterior
+    probabilities of the variables of the current time-slice are used as prior
+    probabilties for the next time-slice. They are set as the conditional
+    probability distribution of the nodes/variables.
+    """
+
+    def __init__(self, dbn):
+        super(PriorFeedbackExact, self).__init__(dbn)
+        self._dbn = copy.deepcopy(dbn)
+        self._ft = None
+
+    def marginal_probabilities(self, variable):
+        """Get the marginal probabilities of a variable.
+
+        See `DBNInferenceMethod.marginal_probabilities`.
+        """ 
+        f = self.marginals([variable])
+        return dict(zip(f.values[variable], list(f.get_potential())))
+
+    def marginals(self, variables):
+        """Get the marginals of a variable.
+
+        Parameters
+        ----------
+        variables : [String,]
+            List containing the names of the variables whose joint prior
+            or posterior marginals are desired.
+            Joint marginals for multiple variables can only be computed
+            if they are contained in any clique.     
+                
+        Returns
+        -------
+        Factor
+            A factor containing the desired marginals.
+        """
+        return self._ft.marginals(variables)
+
+    def _set_evidence(self, evidence=None, soft_posteriors=False, **kwargs):
+        """Set evidence for the network (doing the actual work).
+
+        Parameters
+        ----------
+        evidence : dict
+            See `DBNInferenceMethod.set_evidence`.
+        softPosteriors: bool, optional
+            If softPosterior is set to True, the evidence, should it be an
+            np.array, is interpreted as soft evidence for a desired posterior
+            distribution. See `primo2.exact.FactorTree.set_evidence`.
+        """
+        if self._t == 0:
+            self._ft = exact.FactorTree.create_jointree(self._dbn._b0)
+        elif self._t == 1:
+            self._ft = exact.FactorTree.create_jointree(self._dbn._two_tbn)
+        else:
+            self._ft.reset_factors()
+        self._ft.set_evidence({} if evidence is None else evidence, softPosteriors=soft_posteriors)
+
+    def _unroll(self):
+        """Unroll the network, and proceed in time.
+
+        See `DBNInferenceMethod.unroll`.
+        """
+        for node, node_p in self._dbn._transitions:
+            self._dbn._two_tbn.get_node(node).set_cpd(
+                    self._ft.marginals([node_p]).get_potential())
+        self._t += 1
+
+
+class SoftEvidenceExact(DBNInferenceMethod):
+    """Prior feedback-based approach to unrolling a DBN, using exact inference.
+
+    Inference is exact and based on factor trees. The marginal posterior
+    probabilities of the variables of the current time-slice are used as prior
+    probabilties for the next time-slice. They are set as soft evidence on the
+    FactorTree.
+    """
```

Comment:
Do we need to reference e.g. factorTree.set_evidence for an explanation of what soft evidence means here, or at least what kind of soft evidence is used here?